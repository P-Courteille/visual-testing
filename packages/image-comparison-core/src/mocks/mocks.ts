import type { BeforeScreenshotOptions } from '../helpers/beforeScreenshot.interfaces.js'

export const BEFORE_SCREENSHOT_OPTIONS: BeforeScreenshotOptions = {
    instanceData: {
        appName: 'chrome-app',
        browserName: 'chrome',
        browserVersion: '75.0.1',
        deviceName: '',
        devicePixelRatio: 1,
        initialDevicePixelRatio: 1,
        deviceRectangles: {
            statusBar: { y: 0, x: 0, width: 0, height: 0 },
            homeBar: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 0, width: 0 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
            viewport: { y: 0, x: 0, width: 0, height: 0 },
            bottomBar: { y: 0, x: 0, width: 0, height: 0 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
        },
        isAndroid: false,
        isIOS: false,
        isMobile: false,
        logName: 'chrome-latest',
        name: 'chrome-name',
        nativeWebScreenshot: true,
        platformName: 'Windows 10',
        platformVersion: '1234',
    },
    addressBarShadowPadding: 6,
    disableBlinkingCursor: true,
    disableCSSAnimation: true,
    enableLayoutTesting: false,
    noScrollBars: true,
    toolBarShadowPadding: 6,
    hideElements: [<HTMLElement>(<any>'<div></div>')],
    removeElements: [<HTMLElement>(<any>'<div></div>')],
    waitForFontsLoaded: true,
}
export const CONFIGURABLE = {
    writable: true,
    configurable: true,
}
export const NAVIGATOR_APP_VERSIONS = {
    ANDROID: {
        7: '5.0 (Linux; Android 7.1.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        8: '5.0 (Linux; Android 8.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        9: '5.0 (Linux; Android 9; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        10: '5.0 (Linux; Android 10; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        11: '5.0 (Linux; Android 11; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
    },
    IOS: {
        10: '5.0 (iPhone; CPU iPhone OS 10_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/10.0 Mobile/15E148 Safari/604.1',
        11: '5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1',
        12: '5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1',
        13: '5.0 (iPhone; CPU iPhone OS 13_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Mobile/15E148 Safari/604.1',
        14: '5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1',
        15: '5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1',
    },
    IPADOS: {
        13: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.1 Safari/605.1.15',
        14: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15',
        15: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Safari/605.1.15',
    },
}
export const ANDROID_DEVICES = {
    NEXUS_5X: {
        height: 732,
        width: 412,
        innerHeight: 604,
        innerWidth: 412,
    },
    NEXUS_5X_INNER_HEIGHT: {
        height: 732,
        width: 412,
        innerHeight: 800,
        innerWidth: 412,
    },
    TABLET_WIDTH: {
        height: 768,
        width: 1024,
        innerHeight: 604,
        innerWidth: 412,
    },
}
export const IOS_DEVICES = {
    IPHONE: {
        height: 667,
        width: 375,
        innerHeight: 553,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_X: {
        height: 812,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_HEIGHT: {
        height: 896,
        width: 1024,
        innerHeight: 719,
        innerWidth: 414,
        scrollWidth: 414,
        sideBar: 0,
    },
    IPHONE_11: {
        height: 896,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPAD: {
        height: 1366,
        width: 1024,
        innerHeight: 1292,
        innerWidth: 1024,
        scrollWidth: 1024,
        sideBar: 0,
    },
    IPAD_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 746,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
    IPAD_BIG_SIZE: {
        height: 5432,
        width: 9876,
        innerHeight: 5324,
        innerWidth: 9768,
        scrollWidth: 9768,
        sideBar: 108,
    },
    IPAD_PRO_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 954,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
}

export const BASE_CHECK_OPTIONS = {
    wic: {
        addressBarShadowPadding: 6,
        autoElementScroll: true,
        addIOSBezelCorners: false,
        autoSaveBaseline: false,
        clearFolder: false,
        userBasedFullPageScreenshot: false,
        enableLegacyScreenshotMethod: false,
        formatImageName: '{tag}-{logName}-{width}x{height}-dpr-{dpr}',
        isHybridApp: false,
        savePerInstance: true,
        toolBarShadowPadding: 6,
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        fullPageScrollTimeout: 1500,
        hideScrollBars: true,
        waitForFontsLoaded: true,
        compareOptions: {
            ignoreAlpha: false,
            ignoreAntialiasing: false,
            ignoreColors: false,
            ignoreLess: false,
            ignoreNothing: false,
            rawMisMatchPercentage: false,
            returnAllCompareData: false,
            saveAboveTolerance: 0,
            scaleImagesToSameSize: false,
            blockOutSideBar: false,
            blockOutStatusBar: false,
            blockOutToolBar: false,
            createJsonReportFiles: false,
            diffPixelBoundingBoxProximity: 5,
        },
        tabbableOptions: {
            circle: {
                backgroundColor: 'rgba(255, 0, 0, 0.4)',
                borderColor: 'rgba(255, 0, 0, 1)',
                borderWidth: 1,
                fontColor: 'rgba(0, 0, 0, 1)',
                fontFamily: 'Arial',
                fontSize: 10,
                size: 10,
            },
            line: {
                color: 'rgba(255, 0, 0, 1)',
                width: 1,
            },
        }
    },
    instanceData: {
        appName: 'TestApp',
        browserName: 'Chrome',
        browserVersion: '118.0.0.0',
        deviceName: 'iPhone 14',
        devicePixelRatio: 2,
        deviceRectangles: {
            bottomBar: { y: 800, x: 0, width: 390, height: 0 },
            homeBar: { x: 0, y: 780, width: 390, height: 34 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 844, width: 390 },
            statusBar: { x: 0, y: 0, width: 390, height: 47 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 390, height: 47 },
            viewport: { y: 47, x: 0, width: 390, height: 733 }
        },
        initialDevicePixelRatio: 2,
        isAndroid: false,
        isIOS: true,
        isMobile: true,
        logName: 'test-log',
        name: 'test-device',
        nativeWebScreenshot: false,
        platformName: 'iOS',
        platformVersion: '17.0'
    },
    folders: {
        actualFolder: '/test/actual',
        baselineFolder: '/test/baseline',
        diffFolder: '/test/diff'
    },
    testContext: {
        commandName: 'checkScreen',
        framework: 'vitest',
        parent: 'test suite',
        title: 'test title',
        tag: 'test-tag',
        instanceData: {
            browser: { name: 'Chrome', version: '118.0.0.0' },
            deviceName: 'iPhone 14',
            platform: { name: 'iOS', version: '17.0' },
            app: 'TestApp',
            isMobile: true,
            isAndroid: false,
            isIOS: true
        }
    }
}
