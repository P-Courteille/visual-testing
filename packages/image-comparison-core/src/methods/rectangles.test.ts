import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import { join } from 'node:path'
import type { ChainablePromiseElement } from 'webdriverio'
import { determineElementRectangles, determineScreenRectangles, determineStatusAddressToolBarRectangles, determineIgnoreRegions } from './rectangles.js'
import { IMAGE_STRING } from '../mocks/mocks.js'
import type { ElementRectanglesOptions, ScreenRectanglesOptions, StatusAddressToolBarRectanglesOptions, DeviceRectangles } from './rectangles.interfaces.js'

vi.mock('@wdio/globals', () => ({
    browser: {
        execute: vi.fn(),
    }
}))

vi.mock('@wdio/logger', () => import(join(process.cwd(), '__mocks__', '@wdio/logger')))

describe('rectangles', () => {
    let mockBrowserInstance: WebdriverIO.Browser
    let mockExecute: ReturnType<typeof vi.fn>
    let mockGetElementRect: ReturnType<typeof vi.fn>

    beforeEach(() => {
        mockExecute = vi.fn()
        mockGetElementRect = vi.fn()

        mockBrowserInstance = {
            execute: mockExecute,
            getElementRect: mockGetElementRect
        } as unknown as WebdriverIO.Browser
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    // Base device rectangles object
    const baseDeviceRectangles: DeviceRectangles = {
        bottomBar: { y: 0, x: 0, width: 0, height: 0 },
        homeBar: { y: 0, x: 0, width: 0, height: 0 },
        leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
        rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
        screenSize: { height: 0, width: 0 },
        statusBar: { y: 0, x: 0, width: 0, height: 0 },
        statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
        viewport: { y: 0, x: 0, width: 0, height: 0 },
    }

    // Helper function to create element rectangles options
    const createElementOptions = (overrides: Partial<ElementRectanglesOptions> = {}): ElementRectanglesOptions => ({
        isAndroid: false,
        devicePixelRatio: 2,
        deviceRectangles: baseDeviceRectangles,
        isAndroidNativeWebScreenshot: false,
        innerHeight: 500,
        isIOS: false,
        initialDevicePixelRatio: 2,
        isEmulated: false,
        ...overrides,
    })

    // Helper function to create screen rectangles options
    const createScreenOptions = (overrides: Partial<ScreenRectanglesOptions> = {}): ScreenRectanglesOptions => ({
        innerHeight: 553,
        innerWidth: 375,
        isAndroidNativeWebScreenshot: false,
        isAndroidChromeDriverScreenshot: false,
        isIOS: false,
        devicePixelRatio: 2,
        isLandscape: false,
        initialDevicePixelRatio: 2,
        enableLegacyScreenshotMethod: false,
        isEmulated: false,
        ...overrides,
    })

    // Helper function to create status address toolbar options
    const createStatusAddressToolBarOptions = (overrides: Partial<StatusAddressToolBarRectanglesOptions> = {}): StatusAddressToolBarRectanglesOptions => ({
        blockOutSideBar: false,
        blockOutStatusBar: false,
        blockOutToolBar: false,
        isAndroid: false,
        isAndroidNativeWebScreenshot: false,
        isMobile: false,
        isViewPortScreenshot: false,
        ...overrides,
    })

    // Helper function to create enhanced device rectangles with data
    const createDeviceRectanglesWithData = (overrides: Partial<DeviceRectangles> = {}): DeviceRectangles => ({
        ...baseDeviceRectangles,
        statusBarAndAddressBar: { y: 0, x: 0, width: 1344, height: 320 },
        viewport: { y: 320, x: 0, width: 1344, height: 2601 },
        bottomBar: { y: 2921, x: 0, width: 1344, height: 71 },
        leftSidePadding: { y: 320, x: 0, width: 0, height: 2601 },
        rightSidePadding: { y: 320, x: 1344, width: 0, height: 2601 },
        ...overrides,
    })

    describe('determineElementRectangles', () => {
        it('should determine them for iOS', async () => {
            const options = createElementOptions({
                isIOS: true,
                innerHeight: 678,
                deviceRectangles: {
                    ...baseDeviceRectangles,
                    viewport: { y: 20, x: 30, width: 0, height: 0 },
                },
            })

            mockExecute.mockResolvedValueOnce({
                height: 120,
                width: 120,
                x: 100,
                y: 10,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })

        it('should determine them for Android Native webscreenshot', async () => {
            const options = createElementOptions({
                isAndroid: true,
                devicePixelRatio: 3,
                initialDevicePixelRatio: 3,
                isAndroidNativeWebScreenshot: true,
                innerHeight: 678,
                deviceRectangles: {
                    ...baseDeviceRectangles,
                    viewport: { y: 200, x: 300, width: 0, height: 0 },
                },
            })

            mockExecute.mockResolvedValueOnce({
                height: 300,
                width: 200,
                x: 100,
                y: 10,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })

        it('should determine them for Android ChromeDriver', async () => {
            const options = createElementOptions({
                isAndroid: true,
                devicePixelRatio: 1,
                initialDevicePixelRatio: 1,
                innerHeight: 678,
                deviceRectangles: {
                    ...baseDeviceRectangles,
                    viewport: { y: 200, x: 300, width: 0, height: 0 },
                },
            })

            mockExecute.mockResolvedValueOnce({
                height: 20,
                width: 375,
                x: 0,
                y: 0,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })

        it('should determine them for a desktop browser', async () => {
            const options = createElementOptions({
                innerHeight: 500,
            })

            mockExecute.mockResolvedValueOnce({
                height: 20,
                width: 375,
                x: 12,
                y: 34,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })

        it('should determine them for emulated device', async () => {
            const options = createElementOptions({
                isEmulated: true,
                innerHeight: 600,
                devicePixelRatio: 3,
            })

            mockExecute.mockResolvedValueOnce({
                height: 50,
                width: 200,
                x: 15,
                y: 25,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })

        it('should throw an error when the element height is 0', async () => {
            const options = createElementOptions()

            mockExecute.mockResolvedValueOnce({
                height: 0,
                width: 375,
                x: 12,
                y: 34,
            })

            await expect(determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: { selector: '#elementID' },
            })).rejects.toThrow('The element, with selector "$(#elementID)",is not visible. The dimensions are 375x0')
        })

        it('should throw an error when the element width is 0', async () => {
            const options = createElementOptions()

            mockExecute.mockResolvedValueOnce({
                height: 375,
                width: 0,
                x: 12,
                y: 34,
            })

            await expect(determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: { selector: '#elementID' },
            })).rejects.toThrow('The element, with selector "$(#elementID)",is not visible. The dimensions are 0x375')
        })

        it('should throw an error when the element width is 0 and no element selector is provided', async () => {
            const options = createElementOptions()

            mockExecute.mockResolvedValueOnce({
                height: 375,
                width: 0,
                x: 12,
                y: 34,
            })

            await expect(determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: {},
            })).rejects.toThrow('The element is not visible. The dimensions are 0x375')
        })

        it('should handle Android webview elements', async () => {
            const options = createElementOptions({
                isAndroid: true,
                isAndroidNativeWebScreenshot: true,
                devicePixelRatio: 2,
                deviceRectangles: {
                    ...baseDeviceRectangles,
                    viewport: { y: 100, x: 50, width: 375, height: 667 },
                },
            })

            mockExecute.mockResolvedValueOnce({
                height: 100,
                width: 200,
                x: 50,
                y: 75,
            })

            const result = await determineElementRectangles({
                browserInstance: mockBrowserInstance,
                base64Image: IMAGE_STRING,
                options,
                element: 'webview-element',
            })

            expect(result).toMatchSnapshot()
            expect(mockExecute).toHaveBeenCalled()
        })
    })

    describe('determineScreenRectangles', () => {
        it('should determine them for iOS', async () => {
            const options = createScreenOptions({
                isIOS: true,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for Android ChromeDriver', async () => {
            const options = createScreenOptions({
                isAndroidChromeDriverScreenshot: true,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for Android Native webscreenshot', async () => {
            const options = createScreenOptions({
                isAndroidNativeWebScreenshot: true,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for desktop browser', async () => {
            const options = createScreenOptions({
                innerHeight: 768,
                innerWidth: 1024,
                devicePixelRatio: 1,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them for emulated device', async () => {
            const options = createScreenOptions({
                isEmulated: true,
                devicePixelRatio: 3,
                isLandscape: true,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })

        it('should determine them with legacy screenshot method', async () => {
            const options = createScreenOptions({
                enableLegacyScreenshotMethod: true,
                isIOS: true,
            })

            expect(determineScreenRectangles(IMAGE_STRING, options)).toMatchSnapshot()
        })
    })

    describe('determineStatusAddressToolBarRectangles', () => {
        it('should determine the rectangles with all blockouts enabled', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutSideBar: true,
                blockOutStatusBar: true,
                blockOutToolBar: true,
                isAndroid: true,
                isAndroidNativeWebScreenshot: true,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles with no blockouts', async () => {
            const options = createStatusAddressToolBarOptions()
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles with only status bar blockout', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutStatusBar: true,
                isAndroid: true,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles with only toolbar blockout', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutToolBar: true,
                isAndroid: true,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles with only sidebar blockout', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutSideBar: true,
                isAndroid: true,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles for iOS with blockouts', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutSideBar: true,
                blockOutStatusBar: true,
                blockOutToolBar: true,
                isAndroid: false,
                isAndroidNativeWebScreenshot: false,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles for non-mobile device', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutStatusBar: true,
                isMobile: false,
                isViewPortScreenshot: false,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should determine the rectangles for Android without native web screenshot', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutStatusBar: true,
                blockOutToolBar: true,
                isAndroid: true,
                isAndroidNativeWebScreenshot: false,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = createDeviceRectanglesWithData()

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })

        it('should handle empty device rectangles', async () => {
            const options = createStatusAddressToolBarOptions({
                blockOutStatusBar: true,
                blockOutToolBar: true,
                isAndroid: true,
                isMobile: true,
                isViewPortScreenshot: true,
            })
            const deviceRectangles = baseDeviceRectangles

            expect(determineStatusAddressToolBarRectangles({ deviceRectangles, options })).toMatchSnapshot()
        })
    })

    describe('determineIgnoreRegions', () => {
        it('should handle mixed ignore regions and elements', async () => {
            const mockElement1 = { elementId: 'element1', selector: '#test1' } as WebdriverIO.Element
            const mockElement2 = { elementId: 'element2', selector: '#test2' } as WebdriverIO.Element
            const mockRegion = { x: 10, y: 20, width: 100, height: 150 }

            mockGetElementRect.mockResolvedValueOnce({ x: 50, y: 60, width: 200, height: 250 })
                .mockResolvedValueOnce({ x: 70, y: 80, width: 300, height: 350 })

            const ignores = [mockElement1, mockRegion, mockElement2]
            const result = await determineIgnoreRegions(mockBrowserInstance, ignores)

            expect(result).toEqual([
                { x: 10, y: 20, width: 100, height: 150 },
                { x: 50, y: 60, width: 200, height: 250 },
                { x: 70, y: 80, width: 300, height: 350 }
            ])
            expect(mockGetElementRect).toHaveBeenCalledTimes(2)
        })

        it('should handle only regions', async () => {
            const mockRegions = [
                { x: 10, y: 20, width: 100, height: 150 },
                { x: 30, y: 40, width: 200, height: 250 }
            ]

            const result = await determineIgnoreRegions(mockBrowserInstance, mockRegions)

            expect(result).toEqual(mockRegions)
            expect(mockGetElementRect).not.toHaveBeenCalled()
        })

        it('should handle only elements', async () => {
            const mockElement1 = { elementId: 'element1', selector: '#test1' } as WebdriverIO.Element
            const mockElement2 = { elementId: 'element2', selector: '#test2' } as WebdriverIO.Element

            mockGetElementRect.mockResolvedValueOnce({ x: 50, y: 60, width: 200, height: 250 })
                .mockResolvedValueOnce({ x: 70, y: 80, width: 300, height: 350 })

            const result = await determineIgnoreRegions(mockBrowserInstance, [mockElement1, mockElement2])

            expect(result).toEqual([
                { x: 50, y: 60, width: 200, height: 250 },
                { x: 70, y: 80, width: 300, height: 350 }
            ])
            expect(mockGetElementRect).toHaveBeenCalledTimes(2)
        })

        it('should throw error for invalid ignore regions', async () => {
            const invalidIgnores = [
                'invalid-string',
                { invalid: 'object' },
                123
            ]

            // @ts-expect-error - invalid ignore regions
            await expect(determineIgnoreRegions(mockBrowserInstance, invalidIgnores))
                .rejects.toThrow('Invalid elements or regions')
        })

        it('should round coordinates', async () => {
            const mockElement = { elementId: 'element1', selector: '#test1' } as WebdriverIO.Element
            const mockRegion = { x: 10.7, y: 20.3, width: 100.9, height: 150.1 }

            mockGetElementRect.mockResolvedValueOnce({ x: 50.4, y: 60.8, width: 200.2, height: 250.6 })

            const result = await determineIgnoreRegions(mockBrowserInstance, [mockRegion, mockElement])

            expect(result).toEqual([
                { x: 11, y: 20, width: 101, height: 150 },
                { x: 50, y: 61, width: 200, height: 251 }
            ])
        })

        it('should handle nested element arrays', async () => {
            const mockElement1 = { elementId: 'element1', selector: '#test1' } as WebdriverIO.Element
            const mockElement2 = { elementId: 'element2', selector: '#test2' } as WebdriverIO.Element

            mockGetElementRect.mockResolvedValueOnce({ x: 50, y: 60, width: 200, height: 250 })
                .mockResolvedValueOnce({ x: 70, y: 80, width: 300, height: 350 })

            const result = await determineIgnoreRegions(mockBrowserInstance, [mockElement1, mockElement2])

            expect(result).toEqual([
                { x: 50, y: 60, width: 200, height: 250 },
                { x: 70, y: 80, width: 300, height: 350 }
            ])
            expect(mockGetElementRect).toHaveBeenCalledTimes(2)
        })

        it('should handle chainable promise elements', async () => {
            const chainableElement = Promise.resolve({ elementId: 'element1', selector: '#test1' } as WebdriverIO.Element)

            mockGetElementRect.mockResolvedValueOnce({ x: 50, y: 60, width: 200, height: 250 })

            const result = await determineIgnoreRegions(mockBrowserInstance, [chainableElement as unknown as ChainablePromiseElement])

            expect(result).toEqual([
                { x: 50, y: 60, width: 200, height: 250 }
            ])
            expect(mockGetElementRect).toHaveBeenCalledTimes(1)
        })
    })
})
