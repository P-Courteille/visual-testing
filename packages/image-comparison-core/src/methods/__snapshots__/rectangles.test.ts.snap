// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rectangles > determineElementRectangles > should determine them for Android ChromeDriver 1`] = `
{
  "height": 20,
  "width": 375,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for Android Native webscreenshot 1`] = `
{
  "height": 900,
  "width": 600,
  "x": 1200,
  "y": 630,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for a desktop browser 1`] = `
{
  "height": 40,
  "width": 750,
  "x": 24,
  "y": 68,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for emulated device 1`] = `
{
  "height": 100,
  "width": 400,
  "x": 30,
  "y": 50,
}
`;

exports[`rectangles > determineElementRectangles > should determine them for iOS 1`] = `
{
  "height": 240,
  "width": 240,
  "x": 260,
  "y": 60,
}
`;

exports[`rectangles > determineElementRectangles > should handle Android webview elements 1`] = `
{
  "height": 200,
  "width": 400,
  "x": 200,
  "y": 350,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for Android ChromeDriver 1`] = `
{
  "height": 1106,
  "width": 2732,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for Android Native webscreenshot 1`] = `
{
  "height": 1536,
  "width": 750,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for desktop browser 1`] = `
{
  "height": 768,
  "width": 1024,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for emulated device 1`] = `
{
  "height": 1659,
  "width": 1125,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them for iOS 1`] = `
{
  "height": 1536,
  "width": 2732,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineScreenRectangles > should determine them with legacy screenshot method 1`] = `
{
  "height": 1536,
  "width": 2732,
  "x": 0,
  "y": 0,
}
`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles for Android without native web screenshot 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles for iOS with blockouts 1`] = `
[
  {
    "height": 320,
    "width": 1344,
    "x": 0,
    "y": 0,
  },
  {
    "height": 71,
    "width": 1344,
    "x": 0,
    "y": 2921,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 0,
    "y": 320,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 1344,
    "y": 320,
  },
]
`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles for non-mobile device 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with all blockouts enabled 1`] = `
[
  {
    "height": 320,
    "width": 1344,
    "x": 0,
    "y": 0,
  },
  {
    "height": 71,
    "width": 1344,
    "x": 0,
    "y": 2921,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 0,
    "y": 320,
  },
  {
    "height": 2601,
    "width": 0,
    "x": 1344,
    "y": 320,
  },
]
`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with no blockouts 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with only sidebar blockout 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with only status bar blockout 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should determine the rectangles with only toolbar blockout 1`] = `[]`;

exports[`rectangles > determineStatusAddressToolBarRectangles > should handle empty device rectangles 1`] = `[]`;
